File: 面试题0106-compress-string-lcci/0106.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

from typing import List

class Solution:
    def compressString(self, S: str) -> str:
        if len(S) == 0:
            return ''
        res = ''
        pre, count = S[0], 1
        for i in range(1, len(S)):
            if S[i] == pre:
                count += 1
            else:
                res += (pre + str(count))
                pre, count = S[i], 1
        
        res += (pre + str(count))

        return S if len(res) >= len(S) else res


if __name__ == '__main__':
    sol = Solution()
    print(sol.compressString('aabcccccaaa'))
    print(sol.compressString('abbccd'))

File: shoope/面试题0106-compress-string-lcci/0106.cpp
#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
public:
    string compressString(string S) {
        int len = S.length();
        if (len == 0) return "";
        string res = "";
        char pre = S[0];
        int count = 1;
        for (int i = 1; i < len; ++i) {
            if (S[i] == pre) {
                ++count;
            }
            else {
                res += pre + to_string(count);
                count = 1;
                pre = S[i];
            }
        }
        res += pre + to_string(count);
        return res.length() >= len ? S : res;
    }
};
int main(int argc, char** args) {
    Solution sol = Solution();
    cout << sol.compressString("aabcccccaaa") << endl;
    cout << sol.compressString("abbccd") << endl;

    return 0;
}